contrib/win32/win32compat/ssh-agent/keyagent-request.c

```diff
@@ -268,6 +365,66 @@ process_sign_request(struct sshbuf* request, struct sshbuf* response, struct age
 	int r, request_invalid = 0, success = 0;
 	struct sshkey *key = NULL;
 
+#ifdef ENABLE_PKCS11
+	int i, count = 0, index = 0;;
+	wchar_t sub_name[MAX_KEY_LENGTH];
+	DWORD sub_name_len = MAX_KEY_LENGTH;
+	DWORD pin_len, epin_len, provider_len;
+	char *pin = NULL, *epin = NULL, *provider = NULL;
+	HKEY root = 0, sub = 0, user_root = 0;
+	struct sshkey **keys = NULL;
+
+	pkcs11_init(0);
+
+	if (get_user_root(con, &user_root) != 0 ||
+		RegOpenKeyExW(user_root, SSH_PKCS11_PROVIDERS_ROOT, 0, STANDARD_RIGHTS_READ | KEY_ENUMERATE_SUB_KEYS | KEY_WOW64_64KEY, &root) != 0) {
+		success = 1;
```

Comment:
Won't this cause a `SSH2_AGENT_SIGN_RESPONSE` with an empty signature if the `SSH_PKCS11_PROVIDERS_ROOT` key doesn't exist? Maybe it would make sense to call `is_reg_sub_key_exists()` here.

Reply:
Great catch! The problem here is that if an attempt to sign with a non pkcs11 key happens before any smartcard provider was added - `SSH_PKCS11_PROVIDERS_ROOT` registry key will not exist and as you mentioned - an empty signature will be returned.
This is an elusive bug because once a provider has been added, `SSH_PKCS11_PROVIDERS_ROOT` will always exist from that point onwards (if all providers were removed at a later point, this key won't have any sub keys, but it will still exist), which makes it harder to find during testing.
To address that I suggest calling `RegCreateKeyExW()` which will create the key if it doesn't exist and will open it if it does, which is exactly what we want. If `RegCreateKeyExW()` fails, we know it's not because `SSH_PKCS11_PROVIDERS_ROOT` doesn't exist, hence we can confidently return `success = 0`.
I've tested this both before and after the fix and in both cases, the behavior is what we'd expect given the above.